src/Components/CSharpExtensionSupport.fs

```diff
@@ -0,0 +1,86 @@
+namespace Ionide.VSCode.FSharp
+
+open Fable.Import.VSCode
+open Fable.Import.VSCode.Vscode
+
+module CSharpExtension =
+
+    let private msCSharpExtensionName = "ms-dotnettools.csharp"
+    let private openvsixCSharpExtensionName = "muhammad-sammy.csharp"
+
+    let private resolvedCSharpExtensionName =
+        if env.appName = "Visual Studio Code" then
+            msCSharpExtensionName
+        else
+            openvsixCSharpExtensionName
+
+    let mutable private hasLookedForCSharp = false
+    let mutable private hasCSharp = false
+    let mutable private csharpExtension: Extension<obj> = null
+    let mutable private hasWarned = false
+
+    let private csharpAvailableContext: bool -> unit =
+        let fn = Context.cachedSetter "fsharp.debugger.available"
+
+        fun value ->
+            hasCSharp <- value
+            fn value
+
+    let isCSharpAvailable () = hasCSharp
+
+    let tryFindCSharpExtension () =
+        if not hasLookedForCSharp then
+            match extensions.getExtension resolvedCSharpExtensionName with
+            | None -> csharpAvailableContext false
+            | Some e ->
+                csharpExtension <- e
+                csharpAvailableContext true
+
+            hasLookedForCSharp <- true
+
+        hasCSharp
+
+    let warnAboutMissingCSharpExtension () =
+        if not hasWarned then
+            window.showWarningMessage (
+                $"The C# extension isn't installed, so debugging and some build tools will not be available. Consider installing the C# extension to enable those features.",
+                [| "Install C# Extension" |]
+            )
+            |> Promise.ofThenable
+            |> Promise.bind (fun c ->
+                if c = Some "Install C# Extension" then
+                    commands.executeCommand ("extension.open", [| Some(box resolvedCSharpExtensionName) |])
+                    |> Promise.ofThenable
+                else
+                    Promise.empty)
+            |> Promise.catch (fun e ->
+                printfn $"Error installing C# extension: {Fable.Core.JS.JSON.stringify e}"
+                Promise.empty)
+            |> ignore<Fable.Core.JS.Promise<_>>
+
+            hasWarned <- true
```

Comment:
These two are designed to be cheap to call, so that components that rely on the debugger can match/call them inside processing loops naievely instead of having to be truly 'reactive'